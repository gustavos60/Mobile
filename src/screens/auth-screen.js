import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import {
  View,
  Text,
  Alert,
  StyleSheet,
} from 'react-native';

import { connect } from 'react-redux';

import { StackActions, NavigationActions } from 'react-navigation';

import Button from '../components/button';
import EmailInput from '../components/email-input';
import PasswordInput from '../components/password-input';

import colors from '../colors';

import {
  login,
  clearErrors,
  createAccount,
} from '../redux/profile/actions';

import { notEmpty, validEmail } from '../validators';

export const AuthScreen = ({
  user,
  authError,
  loggingIn,
  navigation,
  loginAction,
  creatingAccount,
  clearErrorsAction,
  createAccountAction,
}) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [canValidate, setCanValidate] = useState(false);

  useEffect(() => {
    if (user && user.token) {
      navigation.dispatch(StackActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName: 'Tab' })],
      }));
    }
  }, [user]);

  useEffect(() => {
    if (authError) {
      Alert.alert('Erro', authError);
      setTimeout(clearErrorsAction, 300);
    }
  }, [authError]);

  const handleAuth = useCallback((action) => () => {
    if (!canValidate) setCanValidate(true);
    if (!notEmpty(password) || !validEmail(email)) {
      Alert.alert('Atenção', 'Preencha corretamente os campos obrigatórios.');
      return;
    }
    action(email, password);
  }, [email, password, canValidate]);

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.label}>
          Reqres.in
        </Text>
      </View>
      <View style={styles.inputsContainer}>
        <EmailInput
          email={email}
          inputTestId="emailInput"
          valid={!canValidate || (validEmail(email))}
          onChangeText={(text) => setEmail(text)}
        />
        <PasswordInput
          password={password}
          inputTestId="passwordInput"
          valid={!canValidate || (notEmpty(password))}
          onChangeText={(text) => setPassword(text)}
        />
      </View>
      <View style={styles.buttonsContainer}>
        <Button
          label="Cadastro"
          testID="signInButton"
          loading={creatingAccount}
          disabled={loggingIn || creatingAccount}
          buttonStyle={{ marginRight: 10 }}
          onPress={handleAuth(createAccountAction)}
        />
        <Button
          label="Login"
          testID="loginButton"
          loading={loggingIn}
          disabled={loggingIn || creatingAccount}
          buttonStyle={{ marginLeft: 10 }}
          onPress={handleAuth(loginAction)}
        />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
  },
  inputsContainer: {
    margin: 20,
    alignItems: 'center',
    justifyContent: 'center',
  },
  header: {
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.primary,
    height: '30%',
  },
  label: {
    color: colors.white,
    fontSize: 36,
    fontWeight: 'bold',
  },
});

AuthScreen.propTypes = {
  loggingIn: PropTypes.bool,
  creatingAccount: PropTypes.bool,
  user: PropTypes.shape({
    email: PropTypes.string,
    token: PropTypes.string,
    password: PropTypes.string,
  }),
  authError: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.object,
  ]),
  loginAction: PropTypes.func.isRequired,
  clearErrorsAction: PropTypes.func.isRequired,
  createAccountAction: PropTypes.func.isRequired,
};

AuthScreen.defaultProps = {
  user: null,
  authError: null,
  loggingIn: false,
  creatingAccount: true,
};

const mapStateToProps = (state) => ({
  user: state.user.user,
  authError: state.user.error,
  loggingIn: state.user.loggingIn,
  creatingAccount: state.user.creatingAccount,
});

const mapDispatchToProps = (dispatch) => ({
  clearErrorsAction: () => dispatch(clearErrors()),
  loginAction: (email, password) => dispatch(login(email, password)),
  createAccountAction: (email, password) => dispatch(createAccount(email, password)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthScreen);
