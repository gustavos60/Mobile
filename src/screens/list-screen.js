import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  View,
  Text,
  FlatList,
  StyleSheet,
} from 'react-native';

import TabHeaders from '../components/tab-header';
import ListItem from '../components/list-item';

const ListScreen = ({
  places,
  navigation,
}) => {
  const keyExtractor = useCallback((item) => item.uri + item.name);

  const renderEmptyList = useCallback(() => {
    const message = 'Não há nada aqui, para buscar por locais vá a tela de Mapa.';
    return (
      <View style={styles.emptyContainer}>
        <Text>
          {message}
        </Text>
      </View>
    );
  }, []);

  return (
    <View style={styles.container}>
      <TabHeaders
        title="Locais"
        testID="listHeader"
        navigation={navigation}
      />
      <FlatList
        contentContainerStyle={{ marginTop: 10 }}
        data={places}
        keyExtractor={keyExtractor}
        ListEmptyComponent={renderEmptyList}
        renderItem={({ item }) => <ListItem place={item} />}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  emptyContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});

ListScreen.navigationOptions = {
  headerShown: false,
};

ListScreen.propTypes = {
  places: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    latitude: PropTypes.number,
    longitude: PropTypes.number,
    uri: PropTypes.string,
  })).isRequired,
};

const mapStateToProps = (state) => ({
  places: state.places.places,
});

export default connect(mapStateToProps)(ListScreen);
