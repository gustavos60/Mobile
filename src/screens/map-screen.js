import React, { useState, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Text,
  Alert,
  StyleSheet,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import { connect } from 'react-redux';
import Geolocation from '@react-native-community/geolocation';

import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import TabHeaders from '../components/tab-header';
import GoogleMaps from '../components/map';

import { fetchPlaces, clearPlacesError } from '../redux/places/actions';
import colors from '../colors';
import Button from '../components/button';

const geoConfig = {
  timeout: 20000,
  maximumAge: 10000,
  enableHighAccuracy: false,
};

let isMounted;

const MapScreen = ({
  error,
  loading,
  navigation,
  clearErrors,
  fetchNearPlaces,
}) => {
  const [location, setLocation] = useState(null);

  useEffect(() => {
    if (error) {
      Alert.alert('Erro', error);
      setTimeout(clearErrors, 300);
    }
  }, [error]);

  const getUserLocation = useCallback(() => {
    Geolocation.getCurrentPosition(
      (position) => {
        if (isMounted) setLocation(position.coords);
      },
      (err) => Alert.alert(err.message),
      geoConfig,
    );
  }, []);

  useEffect(() => {
    getUserLocation();
    isMounted = true;
    return () => {
      isMounted = false;
    };
  }, []);

  useEffect(() => {
    if (location) fetchNearPlaces(location);
  }, [location]);

  const renderMap = useCallback(() => {
    if (!location) return null;
    const region = {
      latitude: location.latitude,
      longitude: location.longitude,
      latitudeDelta: 0.04,
      longitudeDelta: 0.04,
    };
    return <GoogleMaps region={region} />;
  }, [location]);

  const renderRefreshButton = useCallback(() => {
    if (!location) return null;
    return (
      <TouchableOpacity
        activeOpacity={0.7}
        style={styles.refresh}
        disabled={loading}
        onPress={() => fetchNearPlaces(location)}
      >
        {
          loading
            ? <ActivityIndicator size="small" color="white" />
            : <Icon color={colors.white} size={30} name="refresh" />
        }
      </TouchableOpacity>
    );
  }, [location, fetchNearPlaces]);

  const renderRetryLocation = useCallback(() => {
    if (location) return null;
    return (
      <View style={styles.container}>
        <Text style={{ fontSize: 20, marginBottom: 10 }}>
          Não foi possível obter sua localização.
        </Text>
        <Button
          testID="retryLocationButton"
          label="Tentar novamente"
          onPress={getUserLocation}
        />
      </View>
    );
  }, [location, getUserLocation]);

  return (
    <View style={{ flex: 1 }}>
      <TabHeaders
        title="Mapa"
        testID="mapHeader"
        navigation={navigation}
      />
      {renderMap()}
      {renderRetryLocation()}
      {renderRefreshButton()}
    </View>
  );
};

const styles = StyleSheet.create({
  refresh: {
    position: 'absolute',
    right: 20,
    bottom: 30,
    zIndex: 1,
    height: 60,
    width: 60,
    borderRadius: 30,
    backgroundColor: colors.primary,
    justifyContent: 'center',
    alignItems: 'center',
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

MapScreen.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
    PropTypes.string,
  ]),
  clearErrors: PropTypes.func.isRequired,
  fetchNearPlaces: PropTypes.func.isRequired,
};

MapScreen.defaultProps = {
  error: null,
  loading: false,
};

const mapStateToProps = (state) => ({
  loading: state.places.fetchingPlaces,
  error: state.places.fetchPlacesError,
});

const mapDispatchToProps = (dispatch) => ({
  clearErrors: () => dispatch(clearPlacesError()),
  fetchNearPlaces: (location) => dispatch(fetchPlaces(location)),
});

export default connect(mapStateToProps, mapDispatchToProps)(MapScreen);
