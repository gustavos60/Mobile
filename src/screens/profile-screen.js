import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import {
  View,
  Alert,
  ScrollView,
  StyleSheet,
} from 'react-native';

import Button from '../components/button';
import TabHeaders from '../components/tab-header';
import ProfileField from '../components/profile-field';

import {
  clearErrors,
  updateProfile,
} from '../redux/profile/actions';

import colors from '../colors';
import { notEmpty, validEmail } from '../validators';

const ProfileScreen = ({
  user,
  error,
  loading,
  navigation,
  clearErrorsAction,
  updateProfileAction,
}) => {
  const [profile, setProfile] = useState({});
  const [canEdit, setCanEdit] = useState(false);
  const [canValidate, setCanValidate] = useState(false);

  useEffect(() => {
    setProfile(user);
  }, [user]);

  useEffect(() => {
    if (error) {
      Alert.alert('Erro', error);
      setTimeout(clearErrorsAction, 300);
    }
  }, [error, clearErrorsAction]);

  const cancelEdit = useCallback(() => {
    setProfile(user);
    setCanEdit(false);
  }, [user]);

  const updateUser = useCallback(() => {
    if (!profile.email || profile.email.length === 0) {
      if (!canValidate) setCanValidate(true);
      Alert.alert('Atenção', 'Preencha corretamente os campos obrigatórios.');
      return;
    }
    updateProfileAction(profile)
      .then(() => setCanEdit(false));
  }, [profile, canValidate]);

  const renderButtons = useCallback(() => {
    if (!canEdit) {
      return (
        <View style={styles.buttonsContainer}>
          <Button
            testID="profileEditButton"
            label="Editar"
            onPress={() => setCanEdit(true)}
          />
        </View>
      );
    }
    return (
      <View style={styles.buttonsContainer}>
        <Button
          testID="profileCancelButton"
          label="Cancelar"
          disabled={loading}
          buttonStyle={{ backgroundColor: colors.secondary }}
          onPress={cancelEdit}
        />
        <Button
          testID="profileUpdateButton"
          label="Atualizar"
          loading={loading}
          disabled={loading}
          onPress={updateUser}
        />
      </View>
    );
  }, [canEdit, loading, cancelEdit, updateUser]);


  return (
    <View style={{ flex: 1 }}>
      <TabHeaders
        title="Perfil"
        testID="profileHeader"
        navigation={navigation}
      />
      <ScrollView>
        <ProfileField
          testID="profileNameInput"
          label={`Nome${canEdit ? ' *' : ''}`}
          canEdit={canEdit}
          valid={notEmpty(profile.name)}
          placeholder="Nome"
          value={profile.name || ''}
          onChangeText={(name) => setProfile({ ...profile, name })}
        />
        <ProfileField
          testID="profileEmailInput"
          label={`E-mail${canEdit ? ' *' : ''}`}
          canEdit={canEdit}
          placeholder="E-mail"
          valid={validEmail(profile.email)}
          value={profile.email || ''}
          onChangeText={(email) => setProfile({ ...profile, email })}
          textInputProps={{
            autoCompleteType: 'email',
            keyboardType: 'email-address',
          }}
        />
        <ProfileField
          testID="profileJobInput"
          label="Profissão"
          canEdit={canEdit}
          placeholder="Profissão"
          value={profile.job || ''}
          onChangeText={(job) => setProfile({ ...profile, job })}
        />
        <ProfileField
          testID="profileCountryInput"
          label="País"
          canEdit={canEdit}
          placeholder="País"
          value={profile.country || ''}
          onChangeText={(country) => setProfile({ ...profile, country })}
        />
        {renderButtons()}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  buttonsContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-around',
  },
});

ProfileScreen.propTypes = {
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  user: PropTypes.shape({
    job: PropTypes.string,
    name: PropTypes.string,
    email: PropTypes.string,
    country: PropTypes.string,
  }),
  loading: PropTypes.bool,
  clearErrorsAction: PropTypes.func.isRequired,
  updateProfileAction: PropTypes.func.isRequired,
};

ProfileScreen.defaultProps = {
  user: null,
  error: null,
  loading: false,
};

ProfileScreen.navigationOptions = {
  tabBarTestID: 'profileTabIcon',
};

const mapStateToProps = (state) => ({
  user: state.user.user,
  error: state.user.error,
  loading: state.user.updatingProfile,
});

const mapDispatchToProps = (dispatch) => ({
  clearErrorsAction: () => dispatch(clearErrors()),
  updateProfileAction: (body) => dispatch(updateProfile(body)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ProfileScreen);
