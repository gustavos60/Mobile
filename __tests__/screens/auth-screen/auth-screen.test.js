import 'react-native';
import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { AuthScreen } from '../../../src/screens/auth-screen';

const mockedProps = {
  user: null,
  authError: null,
  loggingIn: false,
  creatingAccount: false,
  navigation: {
    dispatch: jest.fn(),
  },
  loginAction: jest.fn(),
  clearErrorsAction: jest.fn(),
  createAccountAction: jest.fn(),
};

describe('AuthScreen tests', () => {
  it('renders correctly', () => {
    const wrapper = renderer.create(<AuthScreen {...mockedProps} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('should disable login button when logging in', () => {
    const wrapper = shallow(<AuthScreen {...mockedProps} loggingIn />);
    const View = wrapper.findWhere((node) => node.prop('testID') === 'loginButton');
    const { disabled } = View.props();
    expect(disabled).toBeTruthy();
  });

  it('should disable signIn button when logging in', () => {
    const wrapper = shallow(<AuthScreen {...mockedProps} loggingIn />);
    const View = wrapper.findWhere((node) => node.prop('testID') === 'signInButton');
    const { disabled } = View.props();
    expect(disabled).toBeTruthy();
  });

  it('should disable login button when creating account', () => {
    const wrapper = shallow(<AuthScreen {...mockedProps} creatingAccount />);
    const View = wrapper.findWhere((node) => node.prop('testID') === 'loginButton');
    const { disabled } = View.props();
    expect(disabled).toBeTruthy();
  });

  it('should disable signIn button when creating account', () => {
    const wrapper = shallow(<AuthScreen {...mockedProps} creatingAccount />);
    const View = wrapper.findWhere((node) => node.prop('testID') === 'signInButton');
    const { disabled } = View.props();
    expect(disabled).toBeTruthy();
  });

  it('should call login action when clicking on login with valid email and password', () => {
    const loginAction = jest.fn();
    const wrapper = mount(<AuthScreen {...mockedProps} loginAction={loginAction} />);
    const emailInput = wrapper.findWhere((node) => node.prop('testID') === 'emailInput').last();
    const passwordInput = wrapper.findWhere((node) => node.prop('testID') === 'passwordInput').last();
    emailInput.props().onChangeText('gustavo@email.com');
    passwordInput.props().onChangeText('123412341');
    const loginButton = wrapper.findWhere((node) => node.prop('testID') === 'loginButton').last();
    loginButton.props().onClick();
    expect(loginAction).toHaveBeenCalled();
  });

  it('should call login action when clicking on login with invalid email', () => {
    const loginAction = jest.fn();
    const wrapper = mount(<AuthScreen {...mockedProps} loginAction={loginAction} />);
    const emailInput = wrapper.findWhere((node) => node.prop('testID') === 'emailInput').last();
    const passwordInput = wrapper.findWhere((node) => node.prop('testID') === 'passwordInput').last();
    emailInput.props().onChangeText('gustavo');
    passwordInput.props().onChangeText('123412341');
    const loginButton = wrapper.findWhere((node) => node.prop('testID') === 'loginButton').last();
    loginButton.props().onClick();
    expect(loginAction).toHaveBeenCalledTimes(0);
  });
});
