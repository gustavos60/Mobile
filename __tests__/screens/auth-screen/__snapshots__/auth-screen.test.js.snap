// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthScreen tests renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#123456",
        "height": "30%",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 36,
          "fontWeight": "bold",
        }
      }
    >
      Reqres.in
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "margin": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderColor": "gray",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "margin": 10,
            "width": 600,
          },
          Object {
            "borderColor": "gray",
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "gray",
              "fontSize": 30,
            },
            Object {
              "marginLeft": 10,
            },
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCompleteType="email"
        keyboardType="email-address"
        onChangeText={[Function]}
        placeholder="Email"
        rejectResponderTermination={true}
        style={
          Object {
            "flex": 1,
            "fontSize": 18,
            "padding": 10,
          }
        }
        testID="emailInput"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "margin": 10,
            "width": 600,
          },
          Object {
            "borderColor": "gray",
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "gray",
              "fontSize": 30,
            },
            Object {
              "marginLeft": 10,
            },
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCompleteType="password"
        onChangeText={[Function]}
        placeholder="Senha"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "flex": 1,
            "fontSize": 18,
            "padding": 10,
          }
        }
        testID="passwordInput"
        underlineColorAndroid="transparent"
        value=""
      />
      <Text
        allowFontScaling={false}
        onPress={[Function]}
        style={
          Array [
            Object {
              "color": "gray",
              "fontSize": 30,
            },
            Object {
              "marginRight": 10,
            },
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        testID="password-toggle-visibility"
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#123456",
          "borderRadius": 30,
          "height": 55,
          "marginRight": 10,
          "opacity": 1,
          "width": 150,
        }
      }
      testID="signInButton"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 16,
              },
              null,
            ]
          }
        >
          Cadastro
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#123456",
          "borderRadius": 30,
          "height": 55,
          "marginLeft": 10,
          "opacity": 1,
          "width": 150,
        }
      }
      testID="loginButton"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 16,
              },
              null,
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;
